Introduce OpenShift as a new source (CRUD operations)

ruda opened the pull request:
Introduce OpenShift as a source for API and CLI tests. These tests are meant for CRUD operations and it adds support for testing with a token (only OpenShift has make use of it). It somehow takes parts of PR #378.

Notes:
* The OpenShift E2E testing (test_openshift.py) requires refactor after this PR is merged.
* Requires Discovery greater than version `1.2` (refer to card DISCOVERY-2.3.5). It will break and produce test failures results for Discover `1.2.5` (current stable) and bellow -- we (IMO) can't filter out tests based on Discovery server versions.

Running test locally against

```
REPOSITORY TAG IMAGE
quay.io/quipucords/quipucords latest df91f12cecff
```

API tests failing. These are already know issues from current and previous version, due the way we run camoyoc not in the same machine as the discovery server.

```
FAILED camayoc/tests/qpc/api/v1/credentials/test_network_creds.py::test_negative_update_to_invalid - assert 'either a password or an ssh_keyfile, not both' in '{"ssh_keyfile":["ssh_keyfile, /sshke...
FAILED camayoc/tests/qpc/api/v1/credentials/test_network_creds.py::test_negative_create_key_and_pass - assert 'either a password or an ssh_keyfile, not both' in '{"ssh_keyfile":["ssh_keyfile, /priva...
================ 2 failed, 288 passed, 35 skipped, 5 warnings in 601.55s (0:10:01) =================
```



CLI tests failing. These tests are already failing for other sources besides openshift and the test for cluster will be address after this PR.

```
FAILED camayoc/tests/qpc/cli/test_openshift.py::test_openshift_clusters[api.discopop.example.com] - camayoc.exceptions.FailedScanException: The scan with ID "6" has failed unexpectedly.
FAILED camayoc/tests/qpc/cli/test_openshift.py::test_openshift_clusters[api.sharedocp4upi412ovn.example.com] - camayoc.exceptions.FailedScanException: The scan with ID "7" has failed unexpectedly.
FAILED camayoc/tests/qpc/cli/test_openshift.py::test_openshift_clusters[api.sharedocp4upi411ovn.example.com] - camayoc.exceptions.FailedScanException: The scan with ID "8" has failed unexpectedly.
FAILED camayoc/tests/qpc/cli/test_openshift.py::test_openshift_clusters[api.sharedocp4upi410ovn.example.com] - camayoc.exceptions.FailedScanException: The scan with ID "9" has failed unexpectedly.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_add_with_ssl_cert_verify_negative[vcenter] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_add_with_ssl_cert_verify_negative[satellite] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_add_with_ssl_cert_verify_negative[openshift] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_add_with_disable_ssl_negative[vcenter] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_add_with_disable_ssl_negative[satellite] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_add_with_disable_ssl_negative[openshift] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_edit_ssl_cert_verify_negative[vcenter] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_edit_ssl_cert_verify_negative[satellite] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_edit_ssl_cert_verify_negative[openshift] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_edit_disable_ssl_negative[vcenter] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_edit_disable_ssl_negative[satellite] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
FAILED camayoc/tests/qpc/cli/test_sources.py::test_edit_disable_ssl_negative[openshift] - pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
================ 16 failed, 316 passed, 9 skipped, 6 warnings in 1138.38s (0:18:58) ================
```

ruda committed:
constants: default port and openshift

Add openshift as another supported sources.
Add the default ports for supported sources.

ruda committed:
API tests: use default ports constant and openshift source

ruda committed:
CLI tests: use default ports constant and openshift source

ruda committed:
qpc_models: add token to the Credential model

Add the authentication token to the Credential model.

Co-authored-by: Nicole Aragao <nicolearagao@gmail.com>

ruda committed:
API tests: add test to create credential with token

mirekdlugosz commented on camayoc/qpc_models.py:
```diff
@@ -168,6 +169,7 @@ def __init__(
         username=None,
         password=None,
         ssh_keyfile=None,
+        token=None,
```
I think changes like that should be accompanied by configuration spec changes right away.

In `camayoc/types/settings.py`, add a new class:

```
class OpenshiftTokenCredentialOptions(BaseModel):
    name: str
    type: Literal["openshift"]
    auth_token: str
```

And a little below:
```
CredentialOptions = Annotated[
    Union[
        PlainNetworkCredentialOptions,
        SSHNetworkCredentialOptions,
        SatelliteCredentialOptions,
        VCenterCredentialOptions,
        OpenshiftTokenCredentialOptions,
    ],
    Field(discriminator="type"),
]
```

I would go with OpenshiftToken because we already know that Openshift username + password are going to be a thing.

mirekdlugosz reviewed (approved):
Left some improvement suggestions.

ruda committed:
API Test: Use DataProvider in test_creds_common

ruda committed:
Consistent nomenclature for Authentication Token

ruda committed:
qpc_models: refactor username and auth_token initialization

ruda committed:
API tests: remove credential tests that have unit test coverage

Tests test_negative_update_to_invalid and
test_negative_update_to_other_type were not implemented!

Also, they are covered by unit tests in Quipucords.

https://github.com/quipucords/quipucords/blob/master/quipucords/api/credential/tests_credential.py

ruda commented:
I may have updated the relevant parts that you call. It looks good to me now.

mirekdlugosz reviewed (changes requested):
I think there might have been misunderstanding, but except for this one place, everything else looks good :+1:

ruda committed:
qpc_models: small refactor

mirekdlugosz reviewed (approved):
Thanks for your patience!
